openmessaging-api/src/main/java/io/openmessaging/api/Client.java

```diff
@@ -0,0 +1,151 @@
+     * Asynchronously create a producer with given {@code ProducerConfiguration} for publishing on a specific topic
+     *
+     * @param topic
+     *            The name of the topic where to produce
+     * @param conf
+     *            The {@code ProducerConfiguration} object
+     * @return Future of the asynchronously created producer object
+     */
+    CompletableFuture<Producer> createProducerAsync(String topic, ProducerConfiguration conf);
+
+    /**
+     * Subscribe to the given topic and subscription combination with given {@code ConsumerConfiguration}
+     *
+     * @param topic
+     *            The name of the topic
+     * @param subscription
+     *            The name of the subscription
+     * @param conf
+     *            The {@code ConsumerConfiguration} object
+     * @return The {@code Consumer} object
+     * @throws OMSException
+     */
+    Consumer subscribe(String topic, String subscription, ConsumerConfiguration conf) throws OMSException;
+
+    /**
+     * Asynchronously subscribe to the given topic and subscription combination using given
+     * {@code ConsumerConfiguration}
+     *
+     * @param topic
+     *            The name of the topic
+     * @param subscription
+     *            The name of the subscription
+     * @param conf
+     *            The {@code ConsumerConfiguration} object
+     * @return Future of the {@code Consumer} object
+     */
+    CompletableFuture<Consumer> subscribeAsync(String topic, String subscription, ConsumerConfiguration conf);
+
+    /**
+     * Create a topic reader with given {@code ReaderConfiguration} for reading messages from the specified topic.
+     * <p>
+     * The Reader provides a low-level abstraction that allows for manual positioning in the topic, without using a
+     * subscription. Reader can only work on non-partitioned topics.
+     * <p>
+     * The initial reader positioning is done by specifying a message id. The options are:
+     * <ul>
+     * <li><code>MessageId.earliest</code> : Start reading from the earliest message available in the topic
+     * <li><code>MessageId.latest</code> : Start reading from the end topic, only getting messages published after the
+     * reader was created
+     * <li><code>MessageId</code> : When passing a particular message id, the reader will position itself on that
+     * specific position. The first message to be read will be the message next to the specified messageId.
+     * </ul>
+     *
+     * @param topic
+     *            The name of the topic where to read
+     * @param startMessageId
+     *            The message id where the reader will position itself. The first message returned will be the one after
+     *            the specified startMessageId
+     * @param conf
+     *            The {@code ReaderConfiguration} object
+     * @return The {@code Reader} object
+     */
+    Reader createReader(String topic, MessageId startMessageId, ReaderConfiguration conf) throws OMSException;
```

Comment:
It seems that `Reader` is a concept from Pulsar?

Reply:
Reader is the `StreamingConsumer` which doesn't maintain any consumption state. it is a sequential `consumer` on `reading` messages. That's why it called `Reader`.

Reply:
-1, the reason as I have listed in the previous discussion. our reinvent reader, future and promise are all not good practice.

Reply:
I am fine with removing Reader for now, since we haven't reached to an agreement about streaming piece. I am fine with starting with simple producer and consumer interface. we can circle back.